pkg/operator/csr/csr_approver.go

```diff
@@ -0,0 +1,180 @@
+		csrName:     csrName,
+		csrApprover: csrApprover,
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithFilteredEventsInformers(factory.NamesFilter(c.csrName), csrInformers.Informer()).
+		ToController("WebhookAuthenticatorCertApprover", eventsRecorder.WithComponentSuffix("webhook-authenticator-cert-approver"))
+}
+
+func (c *csrApproverController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	csr, err := c.csrLister.Get(c.csrName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if approved, denied := getCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+
+	csrCopy := csr.DeepCopy()
+
+	csrPEM, _ := pem.Decode(csr.Spec.Request)
+	if csrPEM == nil {
+		return fmt.Errorf("failed to PEM-parse the CSR block in .spec.request: no CSRs were found")
+	}
+
+	x509CSR, err := x509.ParseCertificateRequest(csrPEM.Bytes)
+	if err != nil {
+		return fmt.Errorf("failed to parse the CSR bytes: %v", err)
+	}
+
+	if x509CSR.Subject.CommonName == csr.Spec.Username {
+		c.denyCSR(ctx, csrCopy, "IllegitimateRequester", "requester cannot request certificates for themselves")
+	}
+
+	csrDecision, denyReason, err := c.csrApprover(csr, x509CSR)
+	if err != nil {
+		return c.denyCSR(ctx, csrCopy, "CSRApprovingFailed", fmt.Sprintf("there was an error during CSR approval: %v", err))
+	}
+
+	switch csrDecision {
+	case CSRDenied:
+		return c.denyCSR(ctx, csrCopy, "CSRDenied", denyReason)
+	case CSRApproved:
+		csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
+			certapiv1.CertificateSigningRequestCondition{
+				Type:    certapiv1.CertificateApproved,
+				Status:  corev1.ConditionTrue,
+				Reason:  "AutoApproved",
+				Message: fmt.Sprintf("Auto-approved CSR %q", c.csrName),
+			},
+		)
+		_, err = c.csrClient.UpdateApproval(ctx, c.csrName, csrCopy, v1.UpdateOptions{})
+		return err
+	case CSRNoOpinion:
+		fallthrough
+	default:
+		return nil
+	}
+}
+
+func (c *csrApproverController) denyCSR(ctx context.Context, csrCopy *certapiv1.CertificateSigningRequest, reason, message string) error {
+	csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
+		certapiv1.CertificateSigningRequestCondition{
+			Type:    certapiv1.CertificateDenied,
+			Status:  corev1.ConditionTrue,
+			Reason:  reason,
+			Message: message,
+		},
+	)
+
+	_, err := c.csrClient.UpdateApproval(ctx, c.csrName, csrCopy, v1.UpdateOptions{})
+	return err
+}
+
+func getCertApprovalCondition(status *certapiv1.CertificateSigningRequestStatus) (approved bool, denied bool) {
+	for _, c := range status.Conditions {
+		if c.Type == certapiv1.CertificateApproved {
+			approved = true
+		}
+		if c.Type == certapiv1.CertificateDenied {
+			denied = true
+		}
+	}
+	return
+}
+
+// ServiceAccountApprover approves CSRs with a given subject issued by the provided service account
+func ServiceAccountApprover(saName, expectedSubject string) (CSRApproveFunc, error) {
```

Comment:
minor, I prefer a struct with these values and an Approve func on it instead of the closure